### ISSUE TITLE
2FA not working

### ISSUE THREAD

**Issue by jimblue** [getgrav/grav-plugin-admin#1338]
Hi,

I’m having trouble with 2FA. Once I've set it up with google authenticator an log out, it's impossible to log in.

I've found that settings are well saved but the secret displayed in admin is different from the one saved in the yaml file.

That's mean the QR code used for setting up google authenticator is not the same as the use by Gav.

The problem is I can't found how to fix this!

On the user page, if I click on save, the 2FA secret doesn’t change in the admin, but it does in the yaml file.

Also when I reload user page, the 2FA secret change in both admin and yaml file but they are never the same.

Let me know if you need more information.

Thanks

**Comment by mahagr** [comment]
Try installing Authy -- I had trouble with Google Authenticator in Android, they were using bad/wrong secrets.

**Comment by jimblue** [comment]
Thank's @mahagr ! I will try but but what you say seem strange. Firstly because I use `Google Authenticator` on IOS. Secondly because I use `Google Authenticator` you so many other system (Github, Gitlab, Digital Ocean etc...) without any issue.

**Comment by jimblue** [comment]
Just tried with `Authy` and the problem is the same.

**Comment by rhukster** [comment]
Are you using Cloudflare or anything like that? I'm simply not able to replicate this.  I've had a few people say the 2FA code is changing on reload, but I can't see how that could even be.  Is your code changing in the account yaml?

**Comment by jimblue** [comment]
Yes I'm using Cloudflare, but it's not working either on my local environment.
And yes as I describe in first post code change in yaml file.

**Comment by rhukster** [comment]
Is it possible to zip up and send me a copy of a site where you see this happening? Then I can test it on my local and isolate if its a code, config, or environment issue.  Cheers!

**Comment by jimblue** [comment]
'right just need to cleanup the project a bit first.
Send you this asap.

**Comment by rhukster** [comment]
It's definitely not an issue with Big Picture skeleton specifically because I just downloaded this to my local, installed admin (using CLI), and the twfa_secret in my just-created user account `admin.yaml` is the exact same one that is shown in the admin profile page.  

I really think you have some weird caching issue or perhaps the wrong yaml file? Because I simply cannot recreate this.

**Comment by m4rkinh0** [comment]
Thank you for your answer.
I just spent 15 minutes playing around.
Since the 2FA did not work, I did turn it off by editing the yaml file (twofa_enable: false) to be able to log in again.
Is that enough ? I am a beginner and learning, just know the basics.

Now the weird thing is: the secret is changing everytime I click on my username -_the QR changes_-, or the save button -_the QR do not change_-, from the admin panel.
I downloaded the yaml file to look inside after each click of the mouse.

(About the QR: this is if I enable again the 2FA in the admin panel, if I leave it disabled I can not see this behaviour but the secret in the yaml file is modified.)

The admin.cache is set to NO.
The global Grav cache if OFF too.

Let me know where I should have a look, or if you prefer I can send you a set of credentials or be on teamviewer... whatever will help us solving the issue.

**Comment by esocode** [comment]
**tl;dr: since 1.4 2FA code is regenerated and saved automatically each time the user page is displayed, thus making 2FA unusable.**

I just encountered the same problem. I have been using 2FA since installing grav 2 years ago but suddenly get the error message "Invalid 2-Factor Authentication code". It seems that the bug has been introduced in 1.4. 

When I log in after disabling 2FA in the YAML file when trying to reactivate 2FA I see that the secret in the YAML and in the UI are different. If I then click on the user name in the menu again the user page updates to show the secret that was the current value in the user YAML. But if I reload the YAML it contains another secret that is not (yet) visible in the UI. Only if I click on the user name again the UI shows the secret from the YAML - but then the YAML has been updated already with the next secret.

If I logout and then use the last secret from the YAML (by manually entering it into Authy) I can login with 2FA again. But If I then just display the user page by clicking on the user name the secret will be updated to a new value in the YAML file (without clicking on _Save_!) while the UI still displays the old value.

**Comment by m4rkinh0** [comment]
@esocode Glad to see I am not alone out there.
@rhukster My offer to let you access my site to see by yourself is still valid.

**Comment by esocode** [comment]
I just upgraded to the latest versions (Grav 1.4.5, Admin 1.8.1) and the problem is now fixed for me.

**Comment by m4rkinh0** [comment]
@esocode Oh thank you. I did update and test. Working now. Great !
@rhukster Sorry and thank you ! (I did not see the changelog for this issue).
